[ClangImporter] Make sure fake locations are always in /some/ buffer.

jrose-apple opened the pull request:
We synthesize fake source locations for module imports that come from Swift code; these locations have to be both valid and distinct for Clang to use. We've mostly been getting away with simply making fake offsets from the main file, but after the offsets start exceeding the size of the buffer they start pointing into some *other* file, and then if Clang's SourceManager tries to order those locations it gets *very* confused.

This commit fixes that by allocating a 256K buffer of zeros and using offsets into that instead. The hope is that a read-only mmap'd buffer of zeros that never gets read (except possibly to look for newlines) will be cheap to allocate.

(Why not just make the main file buffer 256K? Because we actually try to parse that, and there's really no reason for the lexer to go crawl through that file eagerly.)

This test case isn't the best because it doesn't actually fail in the old code, but if only the assertion was added we at least hit that. I did verify with the reproducing project I have that we no longer hit this issue.

rdar://problem/30924269

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Hm, looks like this buffer name appears in diagnostics. Better go back to "\<swift-imported-modules>".

benlangmuir commented on test/ClangImporter/Inputs/many-imports/module.modulemap:
```diff
@@ -0,0 +1,100 @@
+module Module1 { header "module1.h" }
```
Does this test require the modules to be non-empty?  If not, it would be nice to omit the headers.

jrose-apple replied:
Mm. I added the headers in an attempt to make it fail the way it does on the actual project, but it didn't actually work. I'll take them back out.

benlangmuir commented on test/ClangImporter/Inputs/many-imports/module1.h:
```diff
@@ -0,0 +1,2 @@
+// comment
+int x;
```
Beware this test may fail on filesystems that collapse identical files, since clang will then see all of these modules as trying to use the same header.  We've been bitten by that with clang's module tests in the past.

jrose-apple replied:
Oh, that's terrible. Thanks.

jrose-apple committed:
[ClangImporter] Make sure fake locations are always in /some/ buffer.

We synthesize fake source locations for module imports that come from
Swift code; these locations have to be both valid and distinct for
Clang to use. We've mostly been getting away with simply making fake
offsets from the main file, but after the offsets start exceeding the
size of the buffer they start pointing into some /other/ file, and
then if Clang's SourceManager tries to order those locations it gets
/very/ confused.

This commit fixes that by allocating a 256K buffer of zeros and using
offsets into that instead. The hope is that a read-only mmap'd buffer
of zeros that never gets read (except possibly to look for newlines)
will be cheap to allocate.

(Why not just make the main file buffer 256K? Because we actually try
to parse that, and there's really no reason for the lexer to go crawl
through that file eagerly.)

This test case isn't the best because it doesn't actually fail in the
old code, but if only the assertion was added we at least hit that.
I did verify with the reproducing project I have that we no longer
hit this issue.

rdar://problem/30924269

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Hm. @scallanan, @jimingham: TestSwiftGenericStructDebugInfoGenericArray.py [seems to be a flaky test](https://ci.swift.org/job/swift-PR-Linux/6467/).

```
13:20:50 Command invoked: /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/lldb/test/dotest.py --executable /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/lldb-linux-x86_64/bin/lldb --rerun-all-issues -C /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/llvm-linux-x86_64/bin/clang -s 2017-03-24-15_19_14 --results-port 45732 -S fnmac --inferior -p TestSwiftGenericStructDebugInfoGenericArray.py /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/lldb/packages/Python/lldbsuite/test --event-add-entries worker_index=11:int
13:20:50 
13:20:50 Configuration: arch=x86_64 compiler=/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/llvm-linux-x86_64/bin/clang
13:20:50 ----------------------------------------------------------------------
13:20:50 Collected 3 tests
13:20:50 
13:20:50 Stack dump:
13:20:50 0.	HandleCommand(command = "frame var -d run -- arg")
13:20:50 
13:20:50 
[TestSwiftGenericStructDebugInfoGenericArray.py FAILED]
```

@swift-ci Please test Linux
